modules/dcache/src/main/java/org/dcache/pool/statistics/IoStatistics.java

```diff
@@ -0,0 +1,205 @@
+     */
+
+    public void updateWrite(long writeBytes, long writeTime){
+        double writeSpeed = calculateSpeed(writeBytes, writeTime);
+        _writeSpeeds.add(writeSpeed);
+
+        _minWriteSpeed = (_writeRequestNum == 0) ? writeSpeed : Math.min(writeSpeed, _minWriteSpeed);
+        _maxWriteSpeed = (_writeRequestNum == 0) ? writeSpeed : Math.max(writeSpeed, _maxWriteSpeed);
+        _avgWriteSpeed = (_writeRequestNum == 0) ? writeSpeed : calculateNewAvg(_writeRequestNum,_avgWriteSpeed, writeSpeed);
+        _95WriteSpeed = (_writeRequestNum == 0) ? writeSpeed : calculate95Percentile((_writeRequestNum + 1), _writeSpeeds);
+
+        _totalWriteTime += writeTime;
+        _writeRequestNum ++;
+        _writtenBytes += writeBytes;
+    }
+
+    /**
+     *
+     * Calculate the speed of a request.
+     *
+     * @param  bytes
+     *         The number of processed bytes; must be non-negative
+     *
+     * @param  time
+     *         The duration of the request in milliseconds; must be non-negative
+     *
+     * @return The speed of the process in bytes per millisecond.
+     */
+
+    private double calculateSpeed(long bytes, long time){
+        return (double) bytes / (double) time;
+    }
+
+    /**
+     *
+     * Calculate the new average of the speed of all read or write requests.
+     *
+     * @param  requestNum
+     *         The number of read or write requests so far; must be non-negative
+     *
+     * @param  oldAvg
+     *         The old average of the speed of all read or write requests; must be non-negative
+     *
+     * @param  newSpeed
+     *         A new speed value of a read or write request; must be non-negative
+     *
+     * @return The new average of the speed of all read or write requests.
+     */
+
+    private double calculateNewAvg(long requestNum, double oldAvg, double newSpeed){
+        return (requestNum * oldAvg + newSpeed) / (requestNum + 1);
+    }
+
+    /**
+     *
+     * Calculate the 95% percentile of the speed of all read or write requests.
+     *
+     * @param  requestNum
+     *         The number of read or write requests so far; must be non-negative
+     *
+     * @param  speeds
+     *         All speeds of all read or write requests so far.
+     *
+     * @return The calculated 95% percentile of the speed of all read or write requests.
+     */
+
+    private double calculate95Percentile(long requestNum, ArrayList<Double> speeds){
+        Collections.sort(speeds);
+        double index = (double) 0.95 * (double) requestNum;
+
+        if (Math.ceil(index) == index){
+            return (double) 0.5 * (speeds.get((int)index) + speeds.get((int) index + 1));
+        } else {
+            return speeds.get((int)Math.ceil(index));
+        }
+    }
+
+    public long getReadRequestNum() {
+        return _readRequestNum;
+    }
+
+    public long getReadBytes() {
+        return _readBytes;
+    }
+
+    public ArrayList<Double> getReadSpeeds() {
```

Comment:
Sorry I missed this one the first time.

You should return the abstract type rather than the concrete type, i.e., List<Double>.